### PR TITLE
Restructure and update OpenAPI doc

### LINKED ISSUE #33176: Restructure and update OpenAPI docs
### Description

Reorganize the OpenAPI docs for .NET 9 in an attempt to pull together the content for controller-based and minimal API web apps. Also adds/updates some content that was found missing or problematic in a recent user study.

## [Describe endpoints](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/openapi?view=aspnetcore-9.0#describe-endpoints)

OpenAPI supports providing summaries and descriptions of routes that are registered in an app. Minimal APIs support two strategies for setting these properties on a given endpoint, using the following extension methods and attributes:

Summaries: [WithSummary](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.withsummary)
Descriptions: [WithDescription](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.openapiroutehandlerbuilderextensions.withdescription)
Summaries: [EndpointSummaryAttribute](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.endpointsummaryattribute)
Descriptions: [EndpointDescriptionAttribute](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.endpointdescriptionattribute)

**`Get started with Microsoft.AspNetCore.OpenApi`**

## [Customize OpenAPI endpoints with endpoint metadata](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/aspnetcore-openapi?view=aspnetcore-9.0&tabs=visual-studio#customize-openapi-endpoints-with-endpoint-metadata)
The following list shows the endpoint metadata that is used to customize the generated OpenAPI document:

Summaries from [IEndpointSummaryMetadata](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.metadata.iendpointsummarymetadata)
Descriptions from [IEndpointDescriptionMetadata](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.metadata.iendpointdescriptionmetadata)
Request body from [IAcceptsMetadata](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.metadata.iacceptsmetadata)
Response information from [IProducesResponseTypeMetadata](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.metadata.iproducesresponsetypemetadata)
Operation IDs from [IEndpointNameMetadata](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.routing.iendpointnamemetadata)
OpenAPI tags from [ITagsMetadata](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.metadata.itagsmetadata)
To learn more about customizing the generated OpenAPI document by modifying endpoint metadata, see [How to use OpenAPI in Minimal API apps](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/openapi?view=aspnetcore-9.0).

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/openapi?view=aspnetcore-9.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/minimal-apis/openapi.md

### Document ID

b18acd86-50e3-2316-27be-27156651e5a8

### Article author

@Rick-Anderson

---
[Associated WorkItem - 273722](https://dev.azure.com/msft-skilling/Content/_workitems/edit/273722)

### PR DESCRIPTION
**Pull request by mikekistler**
Fixes #33176 

This PR restructures the OpenAPI doc in an attempt to pull together the content for controller-based and minimal API web apps. It also adds/updates some content that was found missing or problematic in our recent user study.

There is still some work to do, mainly in generalizing the request body and responses sections to broaden it to include controller-based apps. I've left some TODO comments in these sections to highlight this.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/minimal-apis/openapi.md](https://github.com/dotnet/AspNetCore.Docs/blob/410d117c27ae6a1c27f8c582fa77e3a448bd2988/aspnetcore/fundamentals/minimal-apis/openapi.md) | [OpenAPI support in minimal API apps](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/openapi?branch=pr-en-us-33157) |
| [aspnetcore/fundamentals/openapi/aspnetcore-openapi.md](https://github.com/dotnet/AspNetCore.Docs/blob/410d117c27ae6a1c27f8c582fa77e3a448bd2988/aspnetcore/fundamentals/openapi/aspnetcore-openapi.md) | [aspnetcore/fundamentals/openapi/aspnetcore-openapi](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/openapi/aspnetcore-openapi?branch=pr-en-us-33157) |
| [aspnetcore/fundamentals/openapi/openapi-tools.md](https://github.com/dotnet/AspNetCore.Docs/blob/410d117c27ae6a1c27f8c582fa77e3a448bd2988/aspnetcore/fundamentals/openapi/openapi-tools.md) | [aspnetcore/fundamentals/openapi/openapi-tools](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/openapi/openapi-tools?branch=pr-en-us-33157) |
| [aspnetcore/fundamentals/openapi/overview.md](https://github.com/dotnet/AspNetCore.Docs/blob/410d117c27ae6a1c27f8c582fa77e3a448bd2988/aspnetcore/fundamentals/openapi/overview.md) | [aspnetcore/fundamentals/openapi/overview](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/openapi/overview?branch=pr-en-us-33157) |
| [aspnetcore/whats-new/dotnet-AspNetCore.Docs-mod5.md](https://github.com/dotnet/AspNetCore.Docs/blob/410d117c27ae6a1c27f8c582fa77e3a448bd2988/aspnetcore/whats-new/dotnet-AspNetCore.Docs-mod5.md) | ["ASP.NET Core docs: What's new for May 2024"](https://review.learn.microsoft.com/en-us/aspnet/core/whats-new/dotnet-AspNetCore.Docs-mod5?branch=pr-en-us-33157) |

### CODE REVIEW

File: aspnetcore/fundamentals/openapi/includes/openapi9.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+:::moniker range=">= aspnetcore-9.0"
+
+The [OpenAPI specification](https://spec.openapis.org/oas/latest.html) is a programming language-agnostic standard for documenting HTTP APIs. This standard is supported in ASP.NET apps through a combination of built-in APIs and open-source libraries. There are three key aspects to OpenAPI integration in an application:
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/openapi/includes/openapi9.md:3]: Should we tell them right away what's supported?

```suggestion
ASP.NET Core supports the generation of OpenAPI documents in controller-based apps and minimal APIs.

The [OpenAPI specification](https://spec.openapis.org/oas/latest.html) is a programming language-agnostic standard for documenting HTTP APIs. This standard is supported in ASP.NET apps through a combination of built-in APIs and open-source libraries. There are three key aspects to OpenAPI integration in an application:
```

**Review by Rick-Anderson — APPROVED** [review]

File: aspnetcore/fundamentals/openapi/includes/openapi9.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+* Generating information about the endpoints in the app.
+* Gathering the information into a format that matches the OpenAPI schema.
+* Exposing the generated OpenAPI schema via a visual UI or a serialized file.
+
+ASP.NET apps provide built-in support for generating information about endpoints in an app via the `Microsoft.AspNetCore.OpenApi` package.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/openapi/includes/openapi9.md:9]:
```suggestion
ASP.NET Core apps provide built-in support for generating information about endpoints in an app via the `Microsoft.AspNetCore.OpenApi` package.
```

File: aspnetcore/fundamentals/openapi/includes/openapi9.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
+:::moniker range=">= aspnetcore-9.0"
+
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/openapi/includes/openapi9.md:2]: Should we tell them right away what's supported?

```suggestion
ASP.NET Core supports the generation of OpenAPI documents in controller-based apps and minimal APIs.
```

File: aspnetcore/fundamentals/openapi/includes/openapi9.md (hunk 1 above)
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/openapi/includes/openapi9.md:3]:
```suggestion
The [OpenAPI specification](https://spec.openapis.org/oas/latest.html) is a programming language-agnostic standard for documenting HTTP APIs. This standard is supported in ASP.NET Core apps through a combination of built-in APIs and open-source libraries. There are three key aspects to OpenAPI integration in an application:
```

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+
+  The `Results<TResult1,TResult2,TResultN>` [union types](https://en.wikipedia.org/wiki/Union_type) declare that a route handler returns multiple `IResult`-implementing concrete types, and any of those types that implement `IEndpointMetadataProvider` will contribute to the endpoint’s metadata.
+
+  The union types implement implicit cast operators. These operators enable the compiler to automatically convert the types specified in the generic arguments to an instance of the union type. This capability has the added benefit of providing compile-time checking that a route handler only returns the results that it declares it does. Attempting to return a type that isn't declared as one of the generic arguments to `Results<TResult1,TResult2,TResultN>` results in a compilation error.
+
+### Excluding endpoints from the generated document
+
+<!-- TODO: Add information for controller-based apps in this section -->
+
+By default, all endpoints that are defined in an app are documented in the generated OpenAPI file. Minimal APIs supports two strategies for excluding a given endpoint from the OpenAPI document, using:
```
Comment (tdykstra) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:342]:
```suggestion
By default, all endpoints that are defined in an app are documented in the generated OpenAPI file. Minimal APIs support two strategies for excluding a given endpoint from the OpenAPI document, using:
```

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+### Customize the OpenAPI endpoint route
+
+By default, the OpenAPI endpoint registered via a call to `MapOpenApi` exposes the document at the `/openapi/{documentName}.json` endpoint. The following code demonstrates how to customize the route at which the OpenAPI document is registered:
+
+```csharp
+app.MapOpenApi("/openapi/{documentName}/openapi.json");
+```
+
+***Note:*** It's possible, but not recommended, to remove the `documentName` route parameter from the endpoint route. When the `documentName` route parameter is removed from the endpoint route, the framework attempts to resolve the document name from the query parameter. Not providing the `documentName` in either the route or query can result in unexpected behavior.
```
Comment (tdykstra) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:404]:
```suggestion
It's possible, but not recommended, to remove the `documentName` route parameter from the endpoint route. When the `documentName` route parameter is removed from the endpoint route, the framework attempts to resolve the document name from the query parameter. Not providing the `documentName` in either the route or query can result in unexpected behavior.
```

**Review by tdykstra — APPROVED** [review]
I like the way you restructured the content. Offline I provided a recommendation for the TOC that would take that in the same direction as what you're already doing here. The suggestions I left in this PR are minor matters.

File: aspnetcore/fundamentals/openapi/includes/openapi9.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ASP.NET Core apps provide built-in support for generating information about endpoints in an app via the `Microsoft.AspNetCore.OpenApi` package.
+
+The following code is generated by the ASP.NET Core minimal web API template and uses OpenAPI:
+
+[!code-csharp[](~/fundamentals/minimal-apis/9.0-samples/WebMinOpenApi/Program.cs?name=snippet_default&highlight=5,9-12)]
+
+In the preceding highlighted code:
+
+* `AddOpenApi` registers services required for OpenAPI document generation into the application's DI container.
+* `MapOpenApi` adds an endpoint into the application for viewing the OpenAPI document serialized into JSON.
```
Comment (wadepickett) [inline aspnetcore/fundamentals/openapi/includes/openapi9.md:18]:
```suggestion
* `MapOpenApi` adds an endpoint into the application for viewing the OpenAPI document serialized into JSON. The OpenAPI endpoint is restricted to the development environment to minimize the risk of exposing sensitive information and reduce the attack surface in production.
```

Comment (wadepickett) [reply aspnetcore/fundamentals/openapi/includes/openapi9.md:18]: I think we should explicitly point out why this is for the development environment only whenever we can.

**Review by wadepickett — APPROVED** [review]
Great work!  Approved, but see my suggestion on explicitly stating why the OpenAPI endpoint would only be made available for the development environment.

File: aspnetcore/fundamentals/openapi/includes/openapi9.md (hunk 6 above)
Comment (tdykstra) [reply aspnetcore/fundamentals/openapi/includes/openapi9.md:18]: AFAIK that applies only to SwaggerUI, not to MapOpenApi

**Comment by Rick-Anderson** [comment]
@captainsafia @JeremyLikness  can one of you review this? We'd like to get it merged and do a few more editorial cleanups.

**Review by captainsafia — APPROVED** [review]
Sorry fo the delay -- I took a look at this earlier but didn't click the big green button. LGTM!

File: aspnetcore/fundamentals/openapi/aspnetcore-openapi.md [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+```
+
+### [Visual Studio Code](#tab/visual-studio-code)
+
+Run the following command from the **Integrated Terminal**:
+
+```dotnetcli
+dotnet add package Microsoft.AspNetCore.OpenApi --prerelease
+```
+
```
Comment (mikekistler) [inline aspnetcore/fundamentals/openapi/aspnetcore-openapi.md:40]:
```suggestion
```

**Comment by mikekistler** [comment]
@Rick-Anderson @tdykstra @captainsafia @wadepickett I pushed a new commit that moves the OpenAPI section up from under Minimal APIs to under APIs, as we discussed. Please have a look to see if I did this right.

**Comment by tdykstra** [comment]
> @Rick-Anderson @tdykstra @captainsafia @wadepickett I pushed a new commit that moves the OpenAPI section up from under Minimal APIs to under APIs, as we discussed. Please have a look to see if I did this right.

It looks like an include file got renamed/moved as a main article file, and a TOC link has .md on the end of the uid.  Can you revert the file rename/move in that commit, and remove .md from the uid line in the TOC, and move the TOC block up to line 787? Or I could make a commit to do that.

**Comment by mikekistler** [comment]
@tdykstra  I removed the .md from the TOC link. Hopefully this is all that is needed.

**Comment by tdykstra** [comment]
@mikekistler The build is failing because of the include file being renamed/moved making it an article file.

**Comment by mikekistler** [comment]
It looks like I can't reuse the Microsoft.dotnet-openapi page because this materializes the new OpenAPI section in prior versions. I'll try to fix this.

**Comment by Rick-Anderson** [comment]
@mikekistler  @tdykstra  is going to fix all the problems

**Pull request merged**